doc/guides/adding-new-napi-api.md

```diff
@@ -0,0 +1,50 @@
+# Contributing a new API to N-API
+
+N-API is Node.js's next generation ABI-stable API for native modules.
+While improving the API surface is encouraged and welcomed, the following are
+a set of principles and guidelines to keep in mind while adding a new
+N-API API.
+
+* A new API **must** adhere to N-API API shape and spirit.
+    * **Must** be a C API.
+    * **Must** not throw exceptions.
+    * **Must** return `napi_status`.
+    * **Should** consume `napi_env`.
+    * **Must** operate only on primitive data types, pointers to primitive
+      datatypes or opaque handles.
+    * **Must** be a necessary API and not a nice to have. Convenience APIs
+      belong in node-addon-api.
+    * **Must** not change the signature of an existing N-API API  or break
+      ABI compatibility with other versions of Node.js.
+* New API **should** be agnostic towards the underlying JavaScript VM.

```

Comment:
Is this a "should" or a "must"?

Reply:
You really need to perform the exercise of stretching the bed sheet onto two different mattresses to see if it really is a one-size-fits-all 🙂

I guess, basically, if this is a "must", then so is implementation on top of multiple engines below.

Reply:
Yeah, also the thinking here was that there might be legitimate reasons why we might want to expose an API that was VM specific (that other VMs would just return an error/silently fail on), and so the thinking was to leave that up to the reviewers of the relevant PR